extensions/default/src/commandsModule.ts

```diff
@@ -0,0 +1,368 @@
+      const {
+        protocol,
+        stage,
+        activeStudyUID,
+      } = hangingProtocolService.getActiveProtocol();
+      const { toggleHangingProtocol } = stateSyncService.getState();
+      const storedHanging = `${activeStudyUID}:${protocolId}:${stageIndex | 0}`;
+      if (
+        protocol.id === protocolId &&
+        (stageIndex === undefined || stageIndex === stage)
+      ) {
+        // Toggling off - restore to previous state
+        const previousState = toggleHangingProtocol[storedHanging] || {
+          protocolId: 'default',
+        };
+        return actions.setHangingProtocol(previousState);
+      } else {
+        stateSyncService.reduce({
+          toggleHangingProtocol: {
+            ...toggleHangingProtocol,
+            [storedHanging]: { protocolId: protocol.id, stageIndex: stage },
+          },
+        });
+        return actions.setHangingProtocol({ protocolId, stageIndex });
+      }
+    },
+
+    deltaStage: ({ direction }) => {
+      const state = viewportGridService.getState();
+      const { hangingProtocolId: protocolId, stageIdx } = state.hpInfo;
+      const { protocol } = hangingProtocolService.getActiveProtocol();
+      for (
+        let stageIndex = stageIdx + direction;
+        stageIndex >= 0 && stageIndex < protocol.stages.length;
+        stageIndex += direction
+      ) {
+        if (protocol.stages[stageIndex].enable !== 'disabled') {
+          return actions.setHangingProtocol({
+            protocolId,
+            stageIndex,
+          });
+        }
+      }
+      uiNotificationService.show({
+        title: 'Change Stage',
+        message: 'The hanging protocol has no more applicable stages',
+        type: 'error',
+        duration: 3000,
+      });
+    },
+
+    previousStage: () => {
+      const state = viewportGridService.getState();
+      const { reuseIdMap } = reuseCachedLayouts(state, stateSyncService);
+      // next stage in hanging protocols
+      hangingProtocolService.previousProtocolStage({ reuseIdMap });
+    },
+
+    /**
+     * Changes the viewport layout in terms of the MxN layout.
+     */
+    setViewportLayout: ({ numRows, numCols }) => {
+      const state = viewportGridService.getState();
+      const { hangingProtocolId, stageIdx } = state.hpInfo;
+
+      const initialInDisplay = [];
+      state.viewports.forEach(vp => {
+        if (vp.displaySetInstanceUIDs) {
+          initialInDisplay.push(...vp.displaySetInstanceUIDs);
+        }
+      });
+
+      // The find or create viewport fills in missing viewports by first
+      // looking for previously used viewports, by position, and secondly
+      // by asking the hanging protocol service to provide a viewport.
+      const findOrCreateViewport = (
+        viewportIdx,
+        positionId,
+        cached,
+        options
+      ) => {
+        const byPositionViewport = cached.byPosition?.[positionId];
+        if (byPositionViewport) return { ...byPositionViewport };
```

Comment:
This is very werid, this function is passed to viewportGridService's setLayout, the setLayout pass in the `viewportIdx` (just please add the index, I hate the inconsistency through the code, you are making some Idx and the rest 99% are index). Then in the first two lines here, based on the passed arguments, it is getting the logic which could have been definied in the viewportGridService itself.